docs/adr/2022-03-23_018-single-state.md

```diff
@@ -0,0 +1,68 @@
+* We are looking at upcoming features to [handle rollbacks](https://github.com/input-output-hk/hydra-poc/issues/185) and dealing with [persisting the head state](https://github.com/input-output-hk/hydra-poc/issues/187).
+  - Both could benefit from the idea, that the `HeadState` is just a result of pure `Event` processing (a.k.a event sourcing).
+  - Right now the `HeadState` kept in `Hydra.Node` alone, is not enough to fully describe the state of the `hydra-node`. Hence it would not be enough to just persist all the `Event`s and replaying them to achieve persistence, nor resetting to some previous `HeadState` in the presence of a rollback.
+
+## Decision
+
+* We define and keep a "blackbox" `ChainState tx` in the `HeadState tx`
+  - It shall not be introspectable to the business logic in `HeadLogic`
+  - It shall contain chain-specific information about the current Hydra Head, which will naturally need to evolve once we have multiple Heads in our feature scope
+
+* We provide the latest `ChainState tx` to `postTx` by extending `PostChainTx tx`
+* We change the callback interface of `Chain` to
+```hs
+type ChainCallback tx m = (ChainState tx -> Maybe (OnChainTx tx)) -> m ()
+```
+with the meaning, that invoking the callback indicates receival of a potential Hydra transaction which is `Maybe` observing a relevant `OnChainTx tx`.
+* We also extend `OnChainTx tx` with a `ChainState tx` to allow for state updates when observing.
+
+## Consequences
+
+* We need to change the construction of `Chain` handles and the call sites of `postTx`
+* We need to change the interface between `Hydra.Chain.Direct.Tx` and `Hydra.Chain.Direct.State` to something else than `OnChainTx` as the former would not have a suitble `ChainState` to construct (see example)
+* We need to extract the state handling (similar to the event queue) out of the `HydraNode` handle and shuffle the main of `hydra-node` a bit to be able to provide the latest `ChainState` to the chain callback as a continuation.
+* We need to make the `ChainState` already serializable (`ToJSON`, `FromJSON`) as it's part of the `PostChainTx` and `OnChainTx` types.
+* We can drop the `TVar` of keeping `OnChainHeadState` in the `Hydra.Chain.Direct` module, which should simplify the code
+* Traces (e.g. `ToPost`) and errors (e.g. `PostTxError`) would include the full `ChainState`, which might be helpful but also possible big
+* We might be able to simplify the `ChainState tx` to be just a `UTxOType tx` later
+
+## To be discussed
+
+* Alternative: We could not extend `PostChainTx` and `OnChainTx` with `ChainState`, but passing it as additional argument to `postTx` and returning a tuple in `ChainCallback`, i.e.
+```hs
+postTx :: MonadThrow m => ChainState tx -> PostChainTx tx -> m ()
+type ChainCallback tx m = (ChainState tx -> Maybe (OnChainTx tx, ChainState tx)) -> m ()
+```
```

Comment:
I do highly prefer the alternative here :point_up: , as it:

1. Keep the base types "simpler"
2. Make more obvious from the type signature what is going on (typical state update `s -> (a, s)`) 
3. Maintain a clearer separation between the stateful part, and the somewhat stateless / deterministic bits which are the chain tx.